File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.max = float('-inf')
        self.count = 0
    def _preOrder(self, node: TreeNode):
        if not node:
            return
        if node.val >= self.max:
            self.max = node.val
            val = node.val
            self.count += 1
        else:
            val = self.max

        self._preOrder(node.left)
        self.max = val
        self._preOrder(node.right)
        
    def goodNodes(self, root: TreeNode) -> int:
        self._preOrder(root)
        return self.count

File: Data Structures & Algorithms/balanced-binary-tree/submission-4.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    private int GetHeight(TreeNode root, ref bool isBalanced){
        if (root == null) return 0;

        int left = GetHeight(root.left, ref isBalanced);
        int right = GetHeight(root.right, ref isBalanced);

        if (isBalanced){
            isBalanced = Math.Abs(left-right) <= 1;
        }

        return 1 + Math.Max(left, right);
    }

    public bool IsBalanced(TreeNode root) {
        bool isBalanced = true;
        GetHeight(root, ref isBalanced);
        return isBalanced;
    }
}
